src/global/Pattern.h

```diff
@@ -171,80 +172,116 @@ class CompactVectorOfStrings {
 
   // Allow serialization via the ad_utility::serialization interface.
   AD_SERIALIZE_FRIEND_FUNCTION(CompactVectorOfStrings) {
-    serializer | arg._data;
-    serializer | arg._offsets;
+    serializer | arg.data_;
+    serializer | arg.offsets_;
   }
 
  private:
-  std::vector<data_type> _data;
-  std::vector<offset_type> _offsets;
+  std::vector<data_type> data_;
+  std::vector<offset_type> offsets_;
 };
 
 namespace detail {
 // Allows the incremental writing of a `CompactVectorOfStrings` directly to a
 // file.
 template <typename data_type>
 struct CompactStringVectorWriter {
-  ad_utility::File _file;
-  off_t _startOfFile;
+ private:
   using offset_type = typename CompactVectorOfStrings<data_type>::offset_type;
-  std::vector<offset_type> _offsets;
-  bool _finished = false;
-  offset_type _nextOffset = 0;
 
+  // The data members are encapsulated in a separate struct to make the
+  // definition of the move-assignment operator easier. NOTE: If you add
+  // additional data members to this class, add them inside the `Data` struct.
+  struct Data {
```

Comment:
I was first confused why it does make the implementation easier, until I realized you have to call `finish()` first. Anyways, we have a class `ad_utility::unique_cleanup::UniqueCleanup` that does something very similar to this, allowing you to define a custom cleanup function to avoid having to manually implement move constructors and assignment operators.
You probably won't be able to get rid of the `Data` wrapper though.

Reply:
But it would call the destructor of the object being overriden by move assignment, wouldn't it? So if you make finish part of the `Data` struct you can just delegate the call to finish to there and it should work or am I missing something?

Reply:
No, when I move one object into another, then no destructor at all is called
(except when I move a variant into another variant that holds a different alternative etc, but not on the outer level)